dialog/file.go

```diff
@@ -0,0 +1,244 @@
+package dialog
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"fyne.io/fyne"
+	"fyne.io/fyne/canvas"
+	"fyne.io/fyne/layout"
+	"fyne.io/fyne/theme"
+	"fyne.io/fyne/widget"
+)
+
+type textWidget interface {
+	fyne.Widget
+	SetText(string)
+}
+
+type fileDialog struct {
+	fileName   textWidget
+	open       *widget.Button
+	breadcrumb *widget.Box
+	files      *fyne.Container
+	parent     fyne.Window
+
+	win      *widget.PopUp
+	current  *fileIcon
+	callback func(string)
+	dir      string
+	save     bool
+}
+
+func (f *fileDialog) makeUI() fyne.CanvasObject {
+	if f.save {
+		saveName := widget.NewEntry()
+		saveName.OnChanged = func(s string) {
+			if s == "" {
+				f.open.Disable()
+			} else {
+				f.open.Enable()
+			}
+		}
+		f.fileName = saveName
+	} else {
+		f.fileName = widget.NewLabel("")
+	}
+
+	label := "Open"
+	if f.save {
+		label = "Save"
+	}
+	f.open = widget.NewButton(label, func() {
+		if f.callback == nil {
+			f.win.Hide()
+			return
+		}
+
+		if f.save {
+			name := f.fileName.(*widget.Entry).Text
+			path := filepath.Join(f.dir, name)
+
+			if _, err := os.Stat(path); os.IsNotExist(err) {
+				f.win.Hide()
+				f.callback(path)
+				return
+			}
+
+			ShowConfirm("Overwrite?", "Are you sure you want to overwrite the file\n"+name+"?",
+				func(ok bool) {
+					if !ok {
+						f.callback("")
+						return
+					}
+
+					f.callback(path)
+					f.win.Hide()
+				}, f.parent)
```

Comment:
Could be better if we check if the existed file is really a file and not a dir. Replacing a dir with file is not common.